examples/hosts/app-integration/schema-upgrade/src/dataTransform.ts

```diff
@@ -0,0 +1,104 @@
+    if (versionTag !== "version" || typeof version !== "string" || version === "") {
+        throw new Error("Can't read version");
+    }
+    return version;
+}
+
+/**
+ * Parse string data in version:one format into an array of simple objects that are easily imported into an
+ * inventory list.
+ * @param stringData - version:one formatted string data
+ * @returns An array of objects, each representing a single inventory item
+ */
+export function parseStringDataVersionOne(stringData: string) {
+    const version = readVersion(stringData);
+    if (version !== "one") {
+        throw new Error(`Expected to parse version one, got version ${version}`);
+    }
+    const itemStrings = stringData.split("\n");
+    itemStrings.shift(); // remove version line
+    return itemStrings.map((itemString) => {
+        const [itemNameString, itemQuantityString] = itemString.split(":");
+        return { name: itemNameString, quantity: parseInt(itemQuantityString, 10) };
+    });
+}
+
+/**
+ * Parse string data in version:two format into an array of simple objects that are easily imported into an
+ * inventory list.
+ * @param stringData - version:two formatted string data
+ * @returns An array of objects, each representing a single inventory item
+ */
+export function parseStringDataVersionTwo(stringData: string) {
+    const version = readVersion(stringData);
+    if (version !== "two") {
+        throw new Error(`Expected to parse version two, got version ${version}`);
+    }
+    const itemStrings = stringData.split("\n");
+    itemStrings.shift(); // remove version line
+    return itemStrings.map((itemString) => {
+        const [itemNameString, itemQuantityString] = itemString.split("\t");
+        return { name: itemNameString, quantity: parseInt(itemQuantityString, 10) };
+    });
+}
+
+function parseStringData(stringData: string) {
+    const version = readVersion(stringData);
+    if (version === "one") {
+        return parseStringDataVersionOne(stringData);
+    } else if (version === "two") {
+        return parseStringDataVersionTwo(stringData);
+    } else {
+        throw new Error(`Don't know how to parse version ${version}`);
+    }
+}
+
+function transformToOne(stringData: string) {
+    const inventoryItems = parseStringData(stringData);
+    const inventoryItemStrings = inventoryItems.map((inventoryItem) => {
+        return `${ inventoryItem.name }:${ inventoryItem.quantity.toString() }`;
+    });
+    return `version:one\n${inventoryItemStrings.join("\n")}`;
+}
+
+function transformToTwo(stringData: string) {
+    const inventoryItems = parseStringData(stringData);
+    const inventoryItemStrings = inventoryItems.map((inventoryItem) => {
+        return `${ inventoryItem.name }\t${ inventoryItem.quantity.toString() }`;
+    });
+    return `version:two\n${inventoryItemStrings.join("\n")}`;
+}
+
+export const inventoryListDataTransformationCallback: DataTransformationCallback = async (
```

Comment:
The idea here is to be able to transform (mostly) every version to every other version (sort of N:N matrix) ? Maybe share some thoughts on how do you think about scaling here, or expressing non-compat variations (for example version 4 can only be derived from versions 1 and 2.

Reply:
I think in the normal case it will be one-way transformation -- here I do it both directions just because the demo lets you click back and forth on 1 vs. 2.  In most normal cases I don't think there's a scenario for backwards conversion.

In fact, I suspect the easiest way for a customer to manage this if they have multiple versions, is to daisy-chain (1->2 only, 2->3 only, 3->4 only) such that they only need to produce exactly one transformer per new format.

Reply:
Added a comment to clarify this in latest